internal/services/serviceprincipals/synchronization_secret_resource.go

```diff
@@ -0,0 +1,217 @@
+package serviceprincipals
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"log"
+	"net/http"
+	"time"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/manicminer/hamilton/msgraph"
+	"github.com/manicminer/hamilton/odata"
+
+	"github.com/hashicorp/terraform-provider-azuread/internal/clients"
+	"github.com/hashicorp/terraform-provider-azuread/internal/helpers"
+	"github.com/hashicorp/terraform-provider-azuread/internal/services/serviceprincipals/parse"
+	"github.com/hashicorp/terraform-provider-azuread/internal/tf"
+	"github.com/hashicorp/terraform-provider-azuread/internal/utils"
+	"github.com/hashicorp/terraform-provider-azuread/internal/validate"
+)
+
+func synchronizationSecretResource() *schema.Resource {
+	return &schema.Resource{
+		CreateContext: synchronizationSecretResourceCreate,
+		ReadContext:   synchronizationSecretResourceRead,
+		UpdateContext: synchronizationSecretResourceUpdate,
+		DeleteContext: synchronizationSecretResourceDelete,
+
+		Timeouts: &schema.ResourceTimeout{
+			Create: schema.DefaultTimeout(5 * time.Minute),
+			Read:   schema.DefaultTimeout(4 * time.Minute),
+			Update: schema.DefaultTimeout(3 * time.Minute),
+			Delete: schema.DefaultTimeout(3 * time.Minute),
+		},
+
+		SchemaVersion: 0,
+
+		Schema: map[string]*schema.Schema{
+			"service_principal_id": {
+				Description:      "The id of the application's service principal for which this synchronization secret should be created",
+				Type:             schema.TypeString,
+				Required:         true,
+				ForceNew:         true,
+				ValidateDiagFunc: validate.UUID,
+			},
+			"id": {
+				Description: "Unique synchronization secret identifier.",
+				Type:        schema.TypeString,
+				Computed:    true,
+			},
+			"credentials": {
```

Comment:
As this is intended to be used as a block, we prefer to use singular rather than plural naming

```suggestion
			"credential": {
```